File: backend/app/schemas/history.py
"""
Pydantic schemas for download history.
"""

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import Field

from app.schemas.base import BaseSchema


class HistoryAction(str, Enum):
    """User action that produced a history entry."""

    SENT_TO_CLIENT = "sent_to_client"
    DOWNLOADED_TORRENT = "downloaded_torrent"


class HistoryStatus(str, Enum):
    """Outcome of a history entry."""

    SUCCESS = "success"
    FAILED = "failed"


class HistorySortBy(str, Enum):
    """Available sort fields for history listing."""

    OCCURRED_AT = "occurred_at"
    TITLE = "title"
    SIZE_BYTES = "size_bytes"


class HistoryEntryCreate(BaseSchema):
    """Payload for logging a client-side download action."""

    action: Literal[HistoryAction.DOWNLOADED_TORRENT] = Field(
        HistoryAction.DOWNLOADED_TORRENT,
        description="Action being recorded; only direct .torrent downloads use this endpoint",
    )
    title: str = Field(..., min_length=1, description="Torrent title")
    size_bytes: int | None = Field(None, ge=0, description="File size in bytes")
    info_url: str | None = Field(None, description="Link to torrent info page")
    torrent_url: str | None = Field(None, description="Direct .torrent download URL")
    magnet_link: str | None = Field(None, description="Magnet URI")
    source_type: Literal["jackett", "prowlarr"] = Field(..., description="Type of indexer instance")
    source_instance_id: int | None = Field(None, description="ID of the source instance")
    source_instance_name: str = Field(..., min_length=1, description="Source instance name")
    indexer: str = Field(..., min_length=1, description="Indexer that produced the result")
    search_query: str | None = Field(
        None, max_length=500, description="Search query that surfaced this result"
    )


class HistoryEntryResponse(BaseSchema):
    """A single history row returned to clients."""

    id: int
    occurred_at: datetime
    action: HistoryAction
    status: HistoryStatus
    title: str
    size_bytes: int | None
    size_formatted: str = Field(..., description="Human-readable file size")
    info_url: str | None
    torrent_url: str | None
    magnet_link: str | None
    source_type: str
    source_instance_id: int | None
    source_instance_name: str
    indexer: str
    client_id: int | None
    client_name: str | None
    search_query: str | None
    error_message: str | None


class HistoryListResponse(BaseSchema):
    """Paginated list of history entries."""

    total: int = Field(..., description="Total number of entries matching the filters")
    limit: int = Field(..., description="Page size used for this response")
    offset: int = Field(..., description="Offset used for this response")
    entries: list[HistoryEntryResponse] = Field(..., description="History entries")


class HistoryLookupItem(BaseSchema):
    """A single item to look up against the download history."""

    title: str = Field(..., min_length=1)
    size_bytes: int | None = Field(None, ge=0)
    info_url: str | None = None


class HistoryLookupRequest(BaseSchema):
    """Batch lookup request."""

    items: list[HistoryLookupItem] = Field(
        ..., max_length=500, description="Items to look up (max 500 per request)"
    )


class HistoryMatchEntry(BaseSchema):
    """A single prior history record matching a search result."""

    id: int
    occurred_at: datetime
    action: HistoryAction
    status: HistoryStatus
    client_name: str | None = None


class HistoryMatch(BaseSchema):
    """All history matches for a single input item."""

    index: int = Field(..., description="Index in the input ``items`` array")
    count: int = Field(..., description="Total number of prior history rows for this item")
    last_occurred_at: datetime = Field(..., description="Timestamp of the most recent prior entry")
    entries: list[HistoryMatchEntry] = Field(
        ...,
        description="Up to 10 most recent matching entries, newest first",
    )


class HistoryLookupResponse(BaseSchema):
    """Lookup response. Only items with at least one match are included."""

    matches: list[HistoryMatch] = Field(..., description="Matches keyed by input index")

File: frontend/src/types/history.ts
import { SourceType } from './search'

export type HistoryAction = 'sent_to_client' | 'downloaded_torrent'
export type HistoryStatus = 'success' | 'failed'
export type HistorySortBy = 'occurred_at' | 'title' | 'size_bytes'
export type SortOrder = 'asc' | 'desc'

export interface HistoryEntry {
  id: number
  occurred_at: string
  action: HistoryAction
  status: HistoryStatus
  title: string
  size_bytes: number | null
  size_formatted: string
  info_url: string | null
  torrent_url: string | null
  magnet_link: string | null
  source_type: SourceType
  source_instance_id: number | null
  source_instance_name: string
  indexer: string
  client_id: number | null
  client_name: string | null
  search_query: string | null
  error_message: string | null
}

export interface HistoryListResponse {
  total: number
  limit: number
  offset: number
  entries: HistoryEntry[]
}

export interface HistoryListParams {
  q?: string
  action?: HistoryAction
  source_type?: SourceType
  source_instance_id?: number
  indexer?: string
  client_id?: number
  status?: HistoryStatus
  since?: string
  until?: string
  min_size_bytes?: number
  max_size_bytes?: number
  sort_by?: HistorySortBy
  sort_order?: SortOrder
  limit?: number
  offset?: number
}

export interface HistoryEntryCreate {
  title: string
  size_bytes?: number | null
  info_url?: string | null
  torrent_url?: string | null
  magnet_link?: string | null
  source_type: SourceType
  source_instance_id?: number | null
  source_instance_name: string
  indexer: string
  search_query?: string | null
}

export interface HistoryLookupItem {
  title: string
  size_bytes?: number | null
  info_url?: string | null
}

export interface HistoryMatchEntry {
  id: number
  occurred_at: string
  action: HistoryAction
  status: HistoryStatus
  client_name: string | null
}

export interface HistoryMatch {
  index: number
  count: number
  last_occurred_at: string
  entries: HistoryMatchEntry[]
}

export interface HistoryLookupResponse {
  matches: HistoryMatch[]
}
